### PR TITLE
Add UTF codec and validations support to the Stdlib

### LINKED ISSUE #10660: Add UTF codecs and validators to the `Stdlib`
Since we have `Uchar.t` UTF encoders in `Buffer`, I have been thinking (for a long time) it would be nice to add UTF decoders to the `Stdlib`. For now I focus on simply providing an API for `Bytes`, the rest can follow later.

I think I finally found a design I'm happy with. It is efficient (no allocations, no exceptions) and can easily be used to devise wasteful higher-level abstractions (e.g [`Uutf`-like folders](https://erratique.ch/software/uutf/doc/Uutf.String.html#1_Stringfolders) or `Seq` based iterators). It also has good properties for best-effort decodes.

Since these things are a pain to tweak and work on with a compiler build on your knees I developed this in [this repository](https://github.com/dbuenzli/stdlib-utf). Note that there's a single [API candidate](https://github.com/dbuenzli/stdlib-utf/blob/master/utf_x.mli) there, just a few different implementations of UTF-8 decoding that I wanted to try, see the [README](https://github.com/dbuenzli/stdlib-utf/blob/master/README.md) for all the details. 

Before I consider making a PR, I'd like to:

1. Know whether upstream wants this.
2. Gather comments and agree on the API (structure and names).
3. Agree on the UTF-8 decoder implementation to use.

For 3. I suggest to use `pat`, it was the fastest on my machine and it lets the compiler nicely works out the dispatch table for us by using character range patterns :-) 

Also I still need to work out a little test suite. 

Feel free to open issues and/or make PR there aswell if you see it fit.

### PR DESCRIPTION
**Pull request by dbuenzli**
This PR adds UTF tools, codecs and validations to the `Uchar`, `Bytes` and `String` modules.

For more information and background discussion see #10660.

Closes #10660.

### CODE REVIEW

**Review by nojb — APPROVED** [review]
LGTM. I looked over the PR and it looks to be in very good shape. I already went over the API in the preliminary discussion and I am happy with the present state, so am approving on that basis. As for the implementation I didn't check every index, but I trust the tests which look comprehensive.

We still need (at least) one other formal review.

**Comment by nojb** [comment]
Off topic remark: now that we have UTF-16 decoders in the stdlib one could imagine doing the UTF-8 <-> UTF-16 conversion needed to interface with native Windows APIs on the OCaml side, which would simplify the C code a bit.

**Comment by dbuenzli** [comment]
Does anybody have an idea about the [problem](https://github.com/ocaml/ocaml/pull/10710/checks?check_run_id=3926934185#step:8:8755) with the manual build ? I tried locally, it fails the same way, but if after that I issue `make` again, it succeeds :-)

**Comment by Octachron** [comment]
The underscores (eg `{2:utf_8` ) in the labels for the documentation sections are the responsible for the issue.

**Comment by dbuenzli** [comment]
> The underscores (eg `{2:utf_8` ) in the labels for the documentation sections are the responsible for the issue.

What do you suggest should be done ? That's valid `ocamldoc` syntax.

**Comment by dbuenzli** [comment]
There seem to be a few tricks [here](https://tex.stackexchange.com/questions/121416/putting-an-underscore-in-a-label) if you tell me where the labels are generated (if they are) I could try to apply them.

**Comment by Octachron** [comment]
I will have a look to see why the manual setting to allow `_` in most context doesn't work here tomorrow.

**Comment by nojb** [comment]
> I will have a look to see why the manual setting to allow `_` in most context doesn't work here tomorrow.

For info, I also had trouble with underscores in #10545 and ended up using a hack https://github.com/ocaml/ocaml/blob/f44236171d937564571eb145fa13ef422ed8f30a/manual/src/library/stdlib-blurb.etex#L80-L81

**Comment by Octachron** [comment]
I have fix for allowing underscore with labels and references while still disabling the latex semantic for `_` outside of math mode in #10713 .

@nojb : it looks like the issue in this case is ocamldoc mangling of the underscore in the generated labels for modules. This can probably be solved with a little more macro glue. I might have a look later.

**Comment by nojb** [comment]
> @nojb : it looks like the issue in this case is ocamldoc mangling of the underscore in the generated labels for modules. This can probably be solved with a little more macro glue. I might have a look later.

Thanks for looking into it!

**Comment by dbuenzli** [comment]
Thanks @Octachron for your swift help. It's all green now. So we need one more good review of all that.

**Comment by Octachron** [comment]
The API looks good to me, I will try to have a look at the bit twiddling soon enough.

File: stdlib/bytes.ml [hunk 1]
```diff
@@ -512,3 +538,289 @@ let set_uint8 = set_int8
...
+      let i = i + 1 in if i > max then dec_invalid 3 else
+      let b3 = get b i in if not_in_x80_to_xBF b3 then dec_invalid 3 else
+      dec_ret 4 (utf_8_uchar_4 b0 b1 b2 b3)
+  | _ -> dec_invalid 1
+
+let set_utf_8_uchar b i u =
+  let set = unsafe_set_uint8 in
+  let max = length b - 1 in
+  match Uchar.to_int u with
+  | u when u < 0 -> assert false
```
Comment (Octachron) [inline stdlib/bytes.ml:642]: I am not sure why only this bound of the Uchar domain is tested? I would expect to either assert false on both side (`u<0` and `u>0x10FFFF`) or not at all?

Comment (dbuenzli) [reply stdlib/bytes.ml:642]: Ah yes it seems I missed adding the check for upper bounds. The code here is supposed to mirror [the code](https://github.com/ocaml/ocaml/blob/c8c3a95a977e582698c3dd6610e271dcdf604066/stdlib/buffer.ml#L137-L213) in `Buffer`.

File: stdlib/bytes.mli [hunk 2]
```diff
@@ -517,6 +517,61 @@ val of_seq : char Seq.t -> t
...
+val set_utf_8_uchar : t -> int -> Uchar.t -> int
+(** [set_utf_8_uchar b i u] UTF-8 encodes [u] at index [i] in [b]
+    and returns the number of bytes [n] that were written starting
+    at [i]. If [n] is [0] there was not enough space to encode [u]
+    at [i] and [b] was left untouched. Otherwise a new character can
+    be encoded at [i + n]. *)
+
+val is_valid_utf_8 : t -> bool
+(** [is_valid_utf_8 b] is [true] if and only if [b] contains valid
+    UTF-8 data. *)
```
Comment (Octachron) [inline stdlib/bytes.mli:539]: I fear that the inattentive reader might read the documentation as:
```
(** [is_valid_utf_8 b] is [true] if and only if [b] only contains valid
    UTF-8 data. *)
```
whereas the function allows the UTF-8 data to end with an error or missing bytes in the middle of a multi-byte sequence. I think that this point should be mentioned.

Comment (dbuenzli) [reply stdlib/bytes.mli:539]: Not sure I understood your comment. I don't think that this: 

> whereas the function allows the UTF-8 data to end with an error or missing bytes in the middle of a multi-byte sequence.

is the case. If it is, it's a bug.

Comment (Octachron) [reply stdlib/bytes.mli:539]: Oh, that would explain why the corresponding UTF-16 functions do not allow a missing low-surrogate. Currently, `is_valid_utf_8` consider that "0xC300" is a valid utf-8 byte sequence which can be tested with
```ocaml
let b = Bytes.create 2
let () = Bytes.set_uint8 b 0 0xC3
let () = Bytes.set_uint8 b 1 0x00
let () = assert (Bytes.is_valid_utf_8 b)
```

File: stdlib/bytes.ml [hunk 3]
```diff
@@ -512,3 +538,292 @@ let set_uint8 = set_int8
...
+  let rec loop max b i =
+    if i > max then true else
+    let get = unsafe_get_uint8 in
+    match Char.unsafe_chr (get b i) with
+    | '\x00' .. '\x7F' -> loop max b (i + 1)
+    | '\xC2' .. '\xDF' ->
+        let last = i + 1 in
+        last > max
+        || not_in_x80_to_xBF (get b last)
+        || loop max b (last + 1)
```
Comment (Octachron) [inline stdlib/bytes.ml]: This sequence of `||` is wrong if the aim is not to stop the validation once we see an invalid continuation byte or the end of the sequence.

Comment (dbuenzli) [reply stdlib/bytes.ml]: Indeed this should be `if last > max then false else …`.

File: stdlib/bytes.mli (hunk 2 above)
Comment (dbuenzli) [reply stdlib/bytes.mli:539]: Indeed. I pushed a fix and I added your test. 

I also now test the `Bytes.is_valid_utf_{8,16be,16le}` functions in the disabled long test in test suite that tests every smallest invalid byte sequences.

File: stdlib/bytes.ml [hunk 4]
```diff
@@ -512,3 +538,299 @@ let set_uint8 = set_int8
...
+   byte is available rather than doing it immediately after the
+   first byte. Contrast with [is_valid_utf_8]. *)
+
+(* UTF-8 *)
+
+let[@inline] not_in_x80_to_xBF b = b lsr 6 <> 0b10
+let[@inline] not_in_xA0_to_xBF b = b lsr 5 <> 0b101
+let[@inline] not_in_x80_to_x9F b = b lsr 5 <> 0b100
+let[@inline] not_in_x90_to_xBF b = b < 0x90 || 0xBF < b
+let[@inline] not_in_x80_to_x8F b = b lsr 4 <> 0x8
```
Comment (Octachron) [inline stdlib/bytes.ml:567]: I wouldn't mind having a comment on the pattern recognized by the test above:
- continuation byte: `0x80..0xBF`
- continuation byte excluding overlong representations: `0xA0..0xBF` and `0x90..0xBF`
- continuation byte excluding high surrogates: `0x80..0x9F`
- continuation byte excluding code point beyond `0x10FFFF`: `0x80..0x8F`

File: stdlib/bytes.ml [hunk 5]
```diff
@@ -512,3 +538,299 @@ let set_uint8 = set_int8
...
+      dec_ret 4 (utf_8_uchar_4 b0 b1 b2 b3)
+  | '\xF4' ->
+      let i = i + 1 in if i > max then dec_invalid 1 else
+      let b1 = get b i in if not_in_x80_to_x8F b1 then dec_invalid 1 else
+      let i = i + 1 in if i > max then dec_invalid 2 else
+      let b2 = get b i in if not_in_x80_to_xBF b2 then dec_invalid 2 else
+      let i = i + 1 in if i > max then dec_invalid 3 else
+      let b3 = get b i in if not_in_x80_to_xBF b3 then dec_invalid 3 else
+      dec_ret 4 (utf_8_uchar_4 b0 b1 b2 b3)
+  | _ -> dec_invalid 1
```
Comment (Octachron) [inline stdlib/bytes.ml:636]: Similarly, adding a comment explaining why `C0`, `C1`, `E0`, `F0`, `F4` and `ED` are isolated in the function above might help the reader to see the link between the pattern matching and utf-8 specification.

**Review by Octachron — APPROVED** [review]
After looking at the implementation, this PR looks great to me. Avoiding code duplication (and potential bugs) for encoding, decoding and validating utf-8 and utf-16 byte sequences is a clear win.

(I have few minor remarks on comments that can be safely ignored.)

File: stdlib/bytes.ml (hunk 4 above)
Comment (dbuenzli) [reply stdlib/bytes.ml:567]: Personally I think mentioning to which part of the Unicode standard these continuations correspond are a distraction from the actual implementation; in the sense it gives the feeling you need to understand these terms to understand the decoding process but that's absolutely not the case (also `overlong representations` are not terms used in TUS).

What really needs to be grasped to understand the implementation and verify its correctness is [table 3.7 on p 124.](http://www.unicode.org/versions/Unicode14.0.0/ch03.pdf) which is mentioned in the comments below.

File: stdlib/bytes.ml (hunk 5 above)
Comment (dbuenzli) [reply stdlib/bytes.ml:636]: Not sure exactly what you wanted to have be explained more. It seems to me that looking at the table in TUS, the pattern match on the first byte and the match bodies is rather clear.

**Comment by dbuenzli** [comment]
Thank you very much @Octachron for your careful review and especially for having found the bug in `Bytes.is_valid_utf_8`. 

Not testing the function on invalid sequences like was done for the decoding ones was a glaring testing gap which is now filled (I ran the long disabled test on my side).

File: stdlib/bytes.ml (hunk 5 above)
Comment (Octachron) [reply stdlib/bytes.ml:636]: I was thinking of people not reading the code and the TUS table synchronously. I cannot find a stable direct link to this table, which might have been an even better option. Maybe the comment `See The Unicode Standard, Table 3.7` could be extended to `This pattern matching is implementing Table 3.7 of The Unicode Standard`?

Anyway, this is a really minor concern.

**Comment by gasche** [comment]
This meets the stdlib standard of two approvals, so it looks like it will be good to merge once the CI passes. Thanks everyone!

**Comment by gasche** [comment]
( @dbuenzli : if you want to fixup things in the git history, now is the time to do it. )

**Comment by dbuenzli** [comment]
I cleaned the patch sequence. Thanks all for taking the time to get this on time before the glaciation.

**Comment by nojb** [comment]
Merged, thanks!

**Pull request merged**